src/backends/torch/torchlib.cc

```diff
@@ -794,190 +853,213 @@ namespace dd
                                              + e.what());
               }
 
-            Tensor loss = tloss.loss(y_pred, y, in_vals);
+            // Compute loss
+            Tensor loss = rank_tloss.loss(y_pred, y, in_vals);
 
             if (iter_size > 1)
               loss /= iter_size;
 
-            double loss_val = loss.item<double>();
-            train_loss += loss_val;
+            // Backward
             loss.backward({},
                           /*retain_graph=*/c10::optional<bool>(retain_graph),
                           /*create_graph=*/false);
-            auto tstop = steady_clock::now();
-            last_it_time
-                += duration_cast<milliseconds>(tstop - tstart).count();
+#pragma omp critical
+            {
+              // Retain loss for statistics
+              double loss_val = loss.item<double>();
+              train_loss += loss_val;
+            }
+          }
 
-            if ((batch_id + 1) % iter_size == 0)
+        // Reduce gradients on device #0
+        auto params = _module.parameters();
+
+        for (size_t j = 1; j < _devices.size(); ++j)
+          {
+            auto params_j = ranks[j].module->parameters();
+
+            for (size_t pi = 0; pi < params.size(); ++pi)
               {
-                tstart = steady_clock::now();
+                Tensor &grad = params[pi].mutable_grad();
+                Tensor gradj = params_j[pi].grad();
+                grad.add_(gradj.to(_main_device));
```

Comment:
Do we need to normalize the gradients by the number of GPUs ?

Reply:
Yes, I can do that